src/util/mmap_impl_win32.h

```diff
@@ -0,0 +1,246 @@
+#ifndef JUMANPP_MMAP_IMPL_WIN32_HPP
+#define JUMANPP_MMAP_IMPL_WIN32_HPP
+
+#define NOMINMAX
+#include <windows.h>
+
+MappedFile::MappedFile(MappedFile &&o) noexcept
+    : fd_{o.fd_},
+      filename_{std::move(o.filename_)},
+      type_{o.type_},
+      size_{o.size_} {
+  o.fd_ = nullptr;
+}
+
+MappedFile &MappedFile::operator=(MappedFile &&o) noexcept {
+  if (this->fd_ != nullptr) {
+    CloseHandle(this->fd_);
+  }
+  fd_ = o.fd_;
+  filename_ = std::move(o.filename_);
+  type_ = o.type_;
+  size_ = o.size_;
+  o.fd_ = nullptr;
+  return *this;
+}
+
+MappedFile::~MappedFile() {
+  if (this->fd_ != nullptr) {
+      CloseHandle(this->fd_);
+  }
+}
+
+Status MappedFile::open(const StringPiece &filename, MMapType type) {
+  if (fd_ != nullptr) {
+    return Status::InvalidState()
+           << "mmap has already opened file " << this->filename_;
+  }
+
+  this->filename_ = filename.str();
+  this->type_ = type;
+
+  DWORD access_mode = 0;
+  DWORD protect_mode = 0;
+  DWORD open_mode = OPEN_EXISTING;
+
+  switch (type) {
+    case MMapType::ReadOnly:
+      access_mode |= GENERIC_READ;
+      protect_mode |= PAGE_READONLY;
+      break;
+    case MMapType::ReadWrite:
+      access_mode |= GENERIC_READ;
+      access_mode |= GENERIC_WRITE;
+      protect_mode |= PAGE_READWRITE;
+      open_mode = CREATE_NEW;
+      break;
+  }
+
+  auto fd = CreateFile(this->filename_.c_str(),
```

Comment:
You should convert the string data to utf-16 here (the design decision is to store everything in utf-8 internally)
https://msdn.microsoft.com/en-us/library/windows/desktop/dd319072(v=vs.85).aspx

```
wchar_t buffer[2048]; //look, no allocs here, ma! paths should be shorter than this most of the time
MultibyteToWideChar(...);
```

Or write a small wrapper which does exactly this.

Reply:
Yes, I totally forgot about converting to wtf-16